### PR TITLE
Use moz.l10n merge instead of compare_locales merge

### PR DESCRIPTION
**Pull request by eemeli**
This changes the merge tool from `compare_locales.merge.merge_channels` to `moz.l10n.resource.add_entries`, and thereby replaces the `compare-locales` parsers & serialisers with `moz.l10n` equivalents.

Besides the changes already in this PR, it'll need to be accompanied by a corresponding update to the data files: moz.l10n includes more parts in the message key, such as the names of each Fluent attribute.

We should also reformat all of the data files at the same time, as otherwise they'll get reformatted one at a time when they're next updated.

PR #63 will need to be merged before this one, as it removes the generally-invalid `click ancestor` string.

### CODE REVIEW

**Review by flodolo — COMMENTED** [review]
I think it would be helpful to see the actual output of the script to evaluate it, at least for the localization files (not storage).

For example, I've only run against master, and while there is a lot of noise (removed empty lines, indentation), this type of change on SelectExpression is surprising.

```diff
 containers-disable-alert-desc =
-    { $tabCount ->
+    { NUMBER($tabCount) ->
         [one] If you disable Container Tabs now, { $tabCount } container tab will be closed. Are you sure you want to disable Container Tabs?
        *[other] If you disable Container Tabs now, { $tabCount } container tabs will be closed. Are you sure you want to disable Container Tabs?
     }
-
 containers-disable-alert-ok-button =
-    { $tabCount ->
+    { NUMBER($tabCount) ->
         [one] Close { $tabCount } Container Tab
        *[other] Close { $tabCount } Container Tabs
     }
```

File: _scripts/requirements.txt [hunk 1]
```diff
@@ -1,3 +1,2 @@
-compare-locales ~= 9.0.2
 moz.l10n ~= 0.1.1
```
[the remark this answers is not in the document]
Comment (eemeli) [reply _scripts/requirements.txt:1]: We've no messages currently impacted by the upstream fix, so there's been no need yet to do a moz.l10n release that includes that fix. I'll do one after mozilla/moz-l10n#14 lands, and can then update here as well.

**Comment by eemeli** [comment]
> I think it would be helpful to see the actual output of the script to evaluate it, at least for the localization files (not storage).

I've held off on including it here because it's a bit tedious to put together, and needs to be redone after each content update: Because of the data format change, each of the active branches needs to be updated locally, and only then pruned.

It would also hide the script change in the middle of hundreds of other changed files.

> For example, I've only run against master, and while there is a lot of noise (removed empty lines, indentation), this type of change on SelectExpression is surprising.

The `$tabCount` → `NUMBER($tabCount)` change is a currently intentional part of the MF2 → Fluent transform. Normally, Fluent relies on e.g. `$tabCount` having a numeric value so that it's implicitly assigned a `FluentNumber` wrapper, which then does category selection. On the other hand, MF2 requires an explicit selector function, assigning the equivalent of `NUMBER()` during the Fluent → MF2 transform if the variant keys look plural-ish.

Not dropping the `NUMBER()` when going back to Fluent makes this duck typing explicit, rather than keeping it internal to Pontoon as it's been so far. If that logic makes a mistake, then at least it's discoverable by looking at the data, rather than just Pontoon.

**Comment by flodolo** [comment]
> It would also hide the script change in the middle of hundreds of other changed files.

My idea was a one-time run on a separate branch (and wip PR).

> The `$tabCount` → `NUMBER($tabCount)` change is a currently intentional part of the MF2 → Fluent transform. Normally, Fluent relies on e.g. `$tabCount` having a numeric value so that it's implicitly assigned a `FluentNumber` wrapper, which then does category selection. On the other hand, MF2 requires an explicit selector function, assigning the equivalent of `NUMBER()` during the Fluent → MF2 transform if the variant keys look plural-ish.

What is the potential impact on other tooling (e.g. linters)? On Pontoon and the output that will end up in l10n repos?

**Comment by flodolo** [comment]
> My idea was a one-time run on a separate branch (and wip PR).

To my surprise, the diff is much smaller when running all branches vs running only master
https://github.com/mozilla-l10n/firefox-l10n-source/compare/test_merge?expand=1

**Comment by eemeli** [comment]
> What is the potential impact on other tooling (e.g. linters)? On Pontoon and the output that will end up in l10n repos?

The `NUMBER()` wrappers should be accepted everywhere, and have no impact on runtime behaviour. Having them in the messages does help ensure that if a value that is presumed to be a number is passed in as a string instead (as in, `'42'` instead of `42`) the message will still select the right category variant. If the value passed in is not numeric at all (as in, `'horse'`), then the default variant will get selected as before.

**Comment by eemeli** [comment]
Once we're happy with the script updates here, we'll need to time the merge of this PR such that:
1. Merge any open `update` PR into `main`.
2. Rebase this PR on `main`.
3. Include a commit on this PR updating the data files for all active branches.
4. Include a commit on this PR applying `l10n-fix` to all localization files.
5. Merge this PR before the next `update` run.

**Comment by flodolo** [comment]
What I'm asking is the effect downstream.

Once we merge the reformat PR, Pontoon will get strings with `NUMBER()`. Nothing should happen to existing translations, since the strings already exist, but is the UI going to deal transparently with `NUMBER()` in a SelectionExpression as it does for `PLATFORM()`? We should verify that.

Also, what happens next to
a) A locale touching another string in the same file. Will all strings be reserialized with `NUMBER()`?
b) A locale translating that string. Will the string be serialized with `NUMBER()`?

**Comment by flodolo** [comment]
> Once we merge the reformat PR, Pontoon will get strings with `NUMBER()`. Nothing should happen to existing translations, since the strings already exist, but is the UI going to deal transparently with `NUMBER()` in a SelectionExpression as it does for `PLATFORM()`? We should verify that.

Looks like we have one in the test project, and it's showing up fine in Pontoon
https://github.com/mozilla-l10n/pontoon-ftl/blob/master/en-US/demo.ftl#L59
https://mozilla-pontoon-staging.herokuapp.com/it/pontoon-test/demo.ftl/?search=number&string=302112

Still curious about the other 2 questions, although they're less important (I assume b is true, not sure about a).

> Also, what happens next to a) A locale touching another string in the same file. Will all strings be reserialized with `NUMBER()`? b) A locale translating that string. Will the string be serialized with `NUMBER()`?

**Review by flodolo — APPROVED** [review]

**Comment by eemeli** [comment]
> Also, what happens next to
> a) A locale touching another string in the same file. Will all strings be reserialized with `NUMBER()`?

No, because `moz.l10n` is not (yet) used to parse & reserialize Fluent messages during Pontoon sync.

> b) A locale translating that string. Will the string be serialized with `NUMBER()`?

Yes, if it had not been localized previously: For an unlocalized message, the structure (including the selector) is generated from the source, and if the source has an explicit `NUMBER()`, then the template will use that as well. If the translated message already has a selector, then that's not modified.

**Review by bcolsson — DISMISSED** [review]

**Review by flodolo — APPROVED** [review]

**Pull request merged**